src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1297 @@
+        
+        if degree is not None:
+            d = degree
+            if d != degree or d < 0:
+                raise ValueError(f"degree argument must be a non-negative integer, got {degree}")
+        else:
+            d = fmpz_mod_poly_degree(self.val, self.ctx.mod.val)
+
+        length = d + 1
+
+        res =  fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = self.ctx
+        fmpz_mod_poly_reverse(res.val, self.val, length, self.ctx.mod.val)
+        return res
+
+    def truncate(self, slong n):
+        r"""
+        Notionally truncate the polynomial to have length ``n``. If 
+        ``n`` is larger than the length of the input, then ``self`` is
+        returned. If ``n`` is not positive, then the zero polynomial 
+        is returned.
+
+        Effectively returns this polynomial :math:`\mod x^n`.
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> f = R([1,2,3])
+            >>> f.truncate(3) is f
+            True
+            >>> f.truncate(2)
+            2*x + 1
+            >>> f.truncate(1)
+            1
+            >>> f.truncate(0)
+            0
+            >>> f.truncate(-1)
+            0
+
+        """
+        cdef fmpz_mod_poly res
+
+        length = fmpz_mod_poly_degree(self.val, self.ctx.mod.val)
+        if n > length:
+            return self
+
+        res =  fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = self.ctx
+
+        if n <= 0:
+            return res
+
+        fmpz_mod_poly_set_trunc(res.val, self.val, n, self.ctx.mod.val)
+        return res
+
+    def is_monic(self):
+        """
+        Return whether this polynomial is monic.
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> x = R.gen()
+            >>> f = x**2 + 5*x + 3
+            >>> f.is_monic()
+            True
+            >>> f = 5*x**2 + x + 3
+            >>> f.is_monic()
+            False
+        """
+        return self.leading_coefficient().is_one()
+
+    def monic(self, check=True):
+        """
+        Return this polynomial divided by its leading coefficient.
+
+        If ``check`` is True, raises ValueError if the leading coefficient 
+        is not invertible modulo N. If ``check`` is False and the leading 
+        coefficient is not invertible, the output is undefined.
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> f = R([1,2,3])
+            >>> f.monic()
+            x^2 + 55*x + 109
+        """
+        cdef fmpz_mod_poly res
+        cdef fmpz_t f
+
+        res =  fmpz_mod_poly.__new__(fmpz_mod_poly)
+        if not check:
+            fmpz_mod_poly_make_monic(
+                res.val, self.val, self.ctx.mod.val
+            )
+        else:
+            fmpz_init(f) 
+            fmpz_mod_poly_make_monic_f(
+                f, res.val, self.val, self.ctx.mod.val
+            )
+            if not fmpz_is_one(f):
+                raise ValueError(f"Leading coefficient is not invertible")
+        res.ctx = self.ctx
+        return res
```

Comment:
`fmpz_clear(f)` is needed here. It needs to be called before raising ValueError and it should also be called if not raising ValueError.